lib/request.js

```diff
@@ -576,19 +576,63 @@ AWS.Request = inherit({
           resp.error.retryable = false;
         });
 
+        var shouldCheckContentLength = false;
+        var expectedLen;
+        if (req.httpRequest.method !== 'HEAD') {
+          expectedLen = parseInt(headers['content-length'], 10);
+        }
+        if (expectedLen !== undefined && !isNaN(expectedLen) && expectedLen >= 0) {
+          shouldCheckContentLength = true;
+          var receivedLen = 0;
+        }
+
+        var checkContentLengthAndEmit = function checkContentLengthAndEmit() {
+          if (shouldCheckContentLength && receivedLen !== expectedLen) {
+            stream.emit('error', AWS.util.error(
+              new Error('Stream content length mismatch. Received ' +
+                receivedLen + ' of ' + expectedLen + ' bytes.'),
+              { code: 'StreamContentLengthMismatch' }
+            ));
+          } else {
+            stream.emit('end')
```

Comment:
At least when using node.js 0.10.x and higher (streams 2+), you'll want to actually call `stream.end()`. This will emit the `unpipe` and `end` events, and clean up some listeners on the source/destination streams that aren't being cleaned when we just emit `end`.

That won't work in node 0.8, but just emitting `end` might suffice there.